### PR TITLE
e2e: infra: setup before e2e run

### PR DESCRIPTION
**Pull request by Tal-or**
The e2e test has many prerequisites before
it can actually test the plugin functionality.
    
This patch intend to provides an automatic (optional) way
for setup the cluster and prepare it for the e2e tests.
    
It does the following:
  1. Deploy a MachineConfig to enable NRI
  2. Deploy KubeletConfig to configure reserved CPUs
  3. Deploy the plugin componenets (DaemonSet, ServiceAccount,
        SecurityContextConstraints, etc.)
    
In addition it's possible to revert the cluster to
its initial state once the test run ends.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

### CODE REVIEW

File: pkg/manifests/manifests.go [hunk 1]
```diff
@@ -72,5 +72,31 @@ func Get() (*Manifests, error) {
 		}
 	}
 
+	// update namespace
```
Comment (ffromani) [inline pkg/manifests/manifests.go:75]: looks obvious to me

File: pkg/manifests/manifests.go [hunk 2]
```diff
@@ -39,18 +40,17 @@ type Manifests struct {
 	DS   appsv1.DaemonSet
 	Role rbacv1.Role
 	RB   rbacv1.RoleBinding
-	SSC  securityv1.SecurityContextConstraints
+	SCC  securityv1.SecurityContextConstraints
 }
 
-func Get() (*Manifests, error) {
+func Get(ns string) (*Manifests, error) {
```
Comment (ffromani) [inline pkg/manifests/manifests.go:46]: what if `ns == ""`? should it be an error?

File: pkg/manifests/manifests.go [hunk 3]
```diff
@@ -72,5 +72,31 @@ func Get() (*Manifests, error) {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name: ns,
+		},
+	}
+	mf.DS.Namespace = ns
+	mf.Role.Namespace = ns
+	mf.RB.Namespace = ns
+	mf.SA.Namespace = ns
+
+	// update SecurityContextConstraints user
```
Comment (ffromani) [inline pkg/manifests/manifests.go:86]: same

File: test/e2e/infrastructure/infrastructure.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"fmt"
+	igntypes "github.com/coreos/ignition/v2/config/v3_2/types"
+	"k8s.io/utils/pointer"
+)
+
+func ()  {
+
+}
```
Comment (ffromani) [inline test/e2e/infrastructure/infrastructure.go]: what is this for?

File: test/e2e/infrastructure/infrastructure.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+			Files: []igntypes.File{
+				{
+					Node: igntypes.Node{
+						Path: "/etc/crio/crio.conf.d/99-nri-enablement.conf",
+					},
+					FileEmbedded1: igntypes.FileEmbedded1{
+						Contents: igntypes.Resource{
+							Source: pointer.String(fmt.Sprintf("%s,%s", defaultIgnitionContentSource, contentBase64)),
+						},
+						Mode: pointer.Int(448),
```
Comment (ffromani) [inline test/e2e/infrastructure/infrastructure.go]: this looks like a weird permission set, what do we need here?

File: test/e2e/e2e_suite_test.go [hunk 6]
```diff
@@ -30,37 +30,79 @@ import (
...
 
 var fixture TestFixture
 
 func TestE2e(t *testing.T) {
 	BeforeSuite(func() {
-		fixture.ctx = context.Background()
+		fixture.Ctx = context.Background()
 		Expect(initClient()).ToNot(HaveOccurred())
 		Expect(initK8SClient()).ToNot(HaveOccurred())
+		Expect(infrastructure.Setup(fixture.Ctx, fixture.Cli, GetNamespaceName())).ToNot(HaveOccurred())
```
Comment (ffromani) [inline test/e2e/e2e_suite_test.go:59]: you may want to start early and annotate expectations with more descriptive messages on error.

File: test/e2e/infrastructure/infrastructure.go [hunk 7]
```diff
@@ -1,56 +1,182 @@
...
 
-func ()  {
+const mcpName = "worker"
+
+//go:embed yamls
+var dir embed.FS
 
+func Setup(ctx context.Context, c client.Client, ns string) error {
+	v, ok := os.LookupEnv("E2E_SETUP")
+	if ok && (v == "TRUE" || v == "true") {
```
Comment (ffromani) [inline test/e2e/infrastructure/infrastructure.go:31]: maybe use `strconv.ParseBool`

File: test/e2e/infrastructure/infrastructure.go [hunk 8]
```diff
@@ -1,56 +1,182 @@
...
-			},
-		},
-	}
-
-	data, err := json.Marshal(ignitionConfig)
+func Teardown(ctx context.Context, c client.Client, ns string) error {
+	v, ok := os.LookupEnv("E2E_TEARDOWN")
+	if ok && (v == "TRUE" || v == "true") {
+		return teardown(ctx, c, ns)
+	}
```
Comment (ffromani) [inline test/e2e/infrastructure/infrastructure.go:42]: same

**Review by ffromani — COMMENTED** [review]
nice, initial review

File: pkg/manifests/manifests.go (hunk 1 above)
Comment (Tal-or) [reply pkg/manifests/manifests.go:75]: This was a note I left for myself and forgot to remove.
Thanks

File: test/e2e/infrastructure/infrastructure.go (hunk 4 above)
Comment (Tal-or) [reply test/e2e/infrastructure/infrastructure.go]: removed on last commit

File: pkg/manifests/manifests.go (hunk 2 above)
Comment (Tal-or) [reply pkg/manifests/manifests.go:46]: I the objects would get created on default ns

Comment (ffromani) [reply pkg/manifests/manifests.go:46]: fair, maybe make sure you have test coverage to document/enforce this

File: test/e2e/infrastructure/infrastructure.go (hunk 5 above)
Comment (Tal-or) [reply test/e2e/infrastructure/infrastructure.go]: This is a standard unix file mode. 
Let's see if we can remove it and use default

File: test/e2e/e2e_suite_test.go (hunk 6 above)
Comment (Tal-or) [reply test/e2e/e2e_suite_test.go:59]: What do you mean by start early?
Actually the `Setup` function is very descriptive and tells exactly what went wrong.

File: test/e2e/infrastructure/infrastructure.go (hunk 7 above)
Comment (Tal-or) [reply test/e2e/infrastructure/infrastructure.go:31]: Yes, good I dea

File: test/e2e/infrastructure/infrastructure.go (hunk 5 above)
Comment (Tal-or) [reply test/e2e/infrastructure/infrastructure.go]: Actually I removed it on latest commit. I should layout the commit history in more organized way, sorry for that.

Comment (ffromani) [reply test/e2e/infrastructure/infrastructure.go]: but in standard unix octal `448` makes not much sense. Anyway, if you fixed differently it's fine

File: test/e2e/e2e_suite_test.go (hunk 6 above)
Comment (ffromani) [reply test/e2e/e2e_suite_test.go:59]: start early in the development cycle, piking up this habit. If the error message is descriptive enough I guess it's fine.

Comment (Tal-or) [reply test/e2e/e2e_suite_test.go:59]: I added another message just to be on the safe side

File: pkg/manifests/manifests.go (hunk 2 above)
Comment (Tal-or) [reply pkg/manifests/manifests.go:46]: Done. thanks

**Pull request merged**